agent/engine/docker_image_manager.go

```diff
@@ -303,21 +307,134 @@ func (imageManager *dockerImageManager) removeUnusedImages(ctx context.Context)
 	imageManager.updateLock.Lock()
 	defer imageManager.updateLock.Unlock()
 
-	imageManager.imageStatesConsideredForDeletion = imageManager.imagesConsiderForDeletion(imageManager.getAllImageStates())
+	imageManager.imageStatesConsideredForDeletion = imageManager.imagesConsiderForDeletion(imageManager.getAllImageStates(), ctx)
 	for i := 0; i < imageManager.numImagesToDelete; i++ {
 		err := imageManager.removeLeastRecentlyUsedImage(ctx)
 		if err != nil {
 			seelog.Infof("End of eligible images for deletion: %v; Still have %d image states being managed", err, len(imageManager.getAllImageStates()))
 			break
 		}
 	}
+	if imageManager.deleteNonECSImagesEnable {
+		// remove nonecs containers
+		imageManager.removeNonECSContainers(ctx)
+		// remove nonecs images
+		imageManager.removeNonECSImages(ctx)
+	}
+}
+
+func (imageManager *dockerImageManager) removeNonECSContainers(ctx context.Context) {
+	nonECSContainerIds := imageManager.nonECSContainersIds(ctx)
+	var nonECSContainerRemoveAvaiableIds []string
+	for _, id := range nonECSContainerIds {
+		response, _ := imageManager.client.InspectContainer(ctx, id, dockerclient.InspectContainerTimeout)
+		if (response.State.Status == "exited") && (time.Now().Sub(response.State.FinishedAt) > imageManager.taskCleanupWaitDuration) {
+			nonECSContainerRemoveAvaiableIds = append(nonECSContainerRemoveAvaiableIds, id)
+		}
+	}
+	for _, id := range nonECSContainerRemoveAvaiableIds {
+		imageManager.client.RemoveContainer(ctx, id, dockerclient.RemoveContainerTimeout)
+	}
+}
+
+func (imageManager *dockerImageManager) nonECSContainersIds(ctx context.Context) []string {
+	ecsContainerIds := imageManager.state.GetAllContainerIDs()
+	var allContainerIds []string
+	listContainersResponse := imageManager.client.ListContainers(ctx, true, dockerclient.ListContainersTimeout)
+	if listContainersResponse.Error != nil {
+		return []string{"listContainersResponse Error"}
+	}
+	for _, dockerID := range listContainersResponse.DockerIDs {
+		allContainerIds = append(allContainerIds, dockerID)
+	}
+	nonECSContainerIds := imageManager.exclude(allContainerIds, ecsContainerIds)
+	return nonECSContainerIds
+}
+
+type ImageWithSize struct {
+	ImageName string
+	Size      int64
+}
+
+func (imageManager *dockerImageManager) removeNonECSImages(ctx context.Context) {
```

Comment:
I thought we were removing only the number of images that were not removed in ECS image cleanup. Is that not true?

Reply:
that's true, NonECS images will be removed after ECS images removal in each cleanup time period.